File: tools/testrunner/internal/infracheck/buildstamp_test.go
package infracheck

import (
	"strings"
	"testing"

	"github.com/ggd/testrunner/internal/testkit"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// TestBuildStampThreadedThroughEveryBuildPath — guard for task #66 / defect
// P0-6(a) of docs/_false-completions.md, failure shape S5 ("fixed, but only in
// this environment").
//
// THE DEFECT. apps/client bakes a build stamp into the bundle and the
// VersionBadge shows it at the bottom of every screen, so a screenshot names
// its build. The stamp used to be computed from `git rev-parse` alone, with a
// `catch { return "dev" }`. That works on a laptop and CANNOT work in an image:
// .dockerignore excludes `.git`, and docker/edge.Dockerfile builds on
// node:22-alpine, which has no git binary. So every image ever built baked the
// word "dev", https://ggd.adms.ai/ showed `dev`, and two different images were
// indistinguishable — which is the only reason the badge exists.
// `grep -rn BUILD_STAMP docker deploy Makefile nginx` returned ZERO hits.
//
// THE DETECTION RECIPE, EXECUTED. The audit's recipe for S5 is "every env var /
// build arg must appear in every one of the build tables". This test is that
// grep: the Dockerfile must accept the arg, and EVERY path that builds the edge
// image must pass it. A stamp threaded through one path and not another
// recreates exactly this bug, one deploy later.
func TestBuildStampThreadedThroughEveryBuildPath(t *testing.T) {
	const argName = "GGD_BUILD_STAMP"

	// 1. The image accepts it as a build arg and exposes it to the client build.
	edge := readRepoFile(t, "docker/edge.Dockerfile")
	require.Contains(t, edge, "ARG "+argName,
		"docker/edge.Dockerfile must declare ARG %s — the image has no .git and no git binary", argName)
	require.Contains(t, edge, "ENV "+argName+"=$"+argName,
		"the ARG must become an ENV, or `pnpm --filter @ggd/client build` never sees it")

	// The ARG/ENV pair has to precede the build RUN, otherwise it is inert.
	argIdx := strings.Index(edge, "ARG "+argName)
	runIdx := strings.Index(edge, `pnpm --filter "@ggd/client" build`)
	require.NotEqual(t, -1, runIdx, "client build RUN not found in docker/edge.Dockerfile")
	assert.Less(t, argIdx, runIdx, "ARG %s must be declared BEFORE the client build runs", argName)

	// 2. Every build path passes it. Adding a new one? Add it here too.
	for _, path := range []string{
		"docker/compose.yaml",        // plain local stack
		"docker/compose.family.yaml", // the family/GCP deploy overlay
		"skaffold.yaml",              // kind + helm (make up / make dev)
		"Makefile",                   // computes it from host git and EXPORTs it
	} {
		assert.Contains(t, readRepoFile(t, path), argName,
			"%s builds or drives the edge image but never mentions %s — "+
				"the badge will read UNSTAMPED-BUILD on that path", path, argName)
	}

	// 3. The two compose files must actually wire it into build.args (mentioning
	//    it in a comment is not wiring). Both interpolate from the host env.
	for _, path := range []string{"docker/compose.yaml", "docker/compose.family.yaml"} {
		assert.Contains(t, readRepoFile(t, path), argName+`: "${`+argName,
			"%s must pass %s as a build arg interpolated from the host environment", path, argName)
	}

	// 4. The Makefile is where the value comes from: it must compute it from git
	//    and export it, since compose/skaffold only interpolate what is exported.
	mk := readRepoFile(t, "Makefile")
	assert.Contains(t, mk, "export "+argName,
		"the Makefile must EXPORT %s — compose and skaffold read it from the environment", argName)
	assert.Contains(t, mk, "git rev-parse --short HEAD",
		"the Makefile must compute the stamp from host git (the image cannot)")

	// 5. And the fallback must stay LOUD. "dev" is what made this invisible for
	//    months: it reads like a deliberate label, so nobody questioned it.
	stamp := readRepoFile(t, "apps/client/dev/buildStamp.ts")
	assert.Contains(t, stamp, `UNSTAMPED = "UNSTAMPED-BUILD"`,
		"an unidentifiable build must be visibly broken, not plausibly labelled")
	assert.NotContains(t, stamp, `return "dev"`,
		"the plausible-looking 'dev' fallback is the regression under test")

	testkit.Cover(t, "infra-build-stamp-arg")
}

File: apps/client/src/buildstamp/buildStamp.test.ts
/**
 * BUILD STAMP RESOLUTION (task #66, defect P0-6(a)).
 *
 * The regression under test, in one sentence: the stamp used to be computed
 * from git ALONE, with `catch { return "dev" }` — and a container has neither
 * `.git` (excluded by .dockerignore) nor a git binary (node:22-alpine), so
 * every image ever built baked the word `dev`. `https://ggd.adms.ai/` showed
 * `dev` at the bottom of every screen and two different images were
 * indistinguishable, which is the only thing the badge is for.
 *
 * Two properties are asserted here, and they are the whole fix:
 *   1. an explicit stamp from the ENVIRONMENT wins over git — that is the
 *      container path, where the host computes the value and threads it in as a
 *      docker build arg;
 *   2. when nothing can identify the build the result is LOUD
 *      (`UNSTAMPED-BUILD`), never a plausible-looking `"dev"`. Plausibility is
 *      what made this survive for months.
 *
 * The wiring half — that every build path actually passes the arg — is guarded
 * in Go, over the real files: tools/testrunner/internal/infracheck/buildstamp_test.go.
 */
import { describe, it, expect } from "vitest";
import { cover } from "@ggd/shared/testkit/cover";
import {
  BUILD_STAMP_ENV_VARS,
  REQUIRE_STAMP_ENV_VAR,
  UNSTAMPED,
  computeBuildStamp,
  resolveBuildStamp,
  unstampedBanner,
  utcDate,
} from "../../dev/buildStamp";

const noGit = (): null => null;
const today = (): string => "2026-07-24";

describe("build stamp resolution (build-stamp-env)", () => {
  it("an explicit env stamp wins — this is the ONLY source a container has", () => {
    cover("build-stamp-env");
    const r = resolveBuildStamp({
      env: { GGD_BUILD_STAMP: "49dca64 2026-07-24" },
      git: () => ({ sha: "aaaaaaa", dirty: false }),
      today,
    });
    expect(r.stamp).toBe("49dca64 2026-07-24");
    expect(r.source).toBe("env");
    expect(r.detail).toBe("GGD_BUILD_STAMP");
  });

  it("accepts VITE_BUILD_STAMP too, but GGD_BUILD_STAMP has priority", () => {
    cover("build-stamp-env");
    expect(BUILD_STAMP_ENV_VARS).toEqual(["GGD_BUILD_STAMP", "VITE_BUILD_STAMP"]);
    expect(
      resolveBuildStamp({ env: { VITE_BUILD_STAMP: "abc1234 2026-07-24" }, git: noGit, today })
        .stamp,
    ).toBe("abc1234 2026-07-24");
    expect(
      resolveBuildStamp({
        env: { GGD_BUILD_STAMP: "win 1", VITE_BUILD_STAMP: "lose 2" },
        git: noGit,
        today,
      }).stamp,
    ).toBe("win 1");
  });

  it("an EMPTY build arg is not a stamp — docker passes '' for an unset arg", () => {
    cover("build-stamp-env");
    // `GGD_BUILD_STAMP: "${GGD_BUILD_STAMP:-}"` in compose yields "" when the
    // host never set it. Treating that as a stamp would bake a blank badge.
    const r = resolveBuildStamp({ env: { GGD_BUILD_STAMP: "   " }, git: noGit, today });
    expect(r.source).toBe("none");
    expect(r.stamp).toBe(UNSTAMPED);
  });

  it("skaffold's `<no value>` is not a stamp either", () => {
    cover("build-stamp-env");
    // MEASURED with a probe image: skaffold does NOT fail on an unset env var
    // in `buildArgs: {K: "{{.K}}"}` — it renders the literal `<no value>` and
    // hands it to docker. Accepting it would put `<no value>` on the badge.
    const r = resolveBuildStamp({ env: { GGD_BUILD_STAMP: "<no value>" }, git: noGit, today });
    expect(r.source).toBe("none");
    expect(r.stamp).toBe(UNSTAMPED);
    // …and it must still lose to a real git head rather than shadowing it.
    expect(
      resolveBuildStamp({
        env: { GGD_BUILD_STAMP: "<no value>" },
        git: () => ({ sha: "49dca64", dirty: false }),
        today,
      }).stamp,
    ).toBe("49dca64 2026-07-24");
  });

  it("falls back to git on a machine that has one, and marks a dirty tree", () => {
    cover("build-stamp-env");
    expect(
      resolveBuildStamp({ env: {}, git: () => ({ sha: "49dca64", dirty: false }), today }).stamp,
    ).toBe("49dca64 2026-07-24");
    // Two images off the SAME commit with different working trees must not be
    // one string — the owner builds from a modified tree routinely.
    expect(
      resolveBuildStamp({ env: {}, git: () => ({ sha: "49dca64", dirty: true }), today }).stamp,
    ).toBe("49dca64-dirty 2026-07-24");
  });

  it("with no env and no git the stamp is LOUD, never a plausible 'dev'", () => {
    cover("build-stamp-env");
    const r = resolveBuildStamp({ env: {}, git: noGit, today });
    expect(r.stamp).toBe("UNSTAMPED-BUILD");
    expect(r.source).toBe("none");
    // The exact regression: the old code returned "dev" here, which reads like a
    // deliberate label and so was never questioned.
    expect(r.stamp).not.toBe("dev");
    expect(r.stamp).toMatch(/UNSTAMPED/);
  });

  it("the failure banner names the reason and the fix", () => {
    cover("build-stamp-env");
    const banner = unstampedBanner(resolveBuildStamp({ env: {}, git: noGit, today }));
    expect(banner).toContain("BUILD STAMP MISSING");
    expect(banner).toContain("GGD_BUILD_STAMP");
    expect(banner).toContain(UNSTAMPED);
  });

  it("computeBuildStamp logs its provenance, and can be made fatal", () => {
    cover("build-stamp-env");
    const lines: string[] = [];
    const stamp = computeBuildStamp(
      { GGD_BUILD_STAMP: "49dca64 2026-07-24" },
      (m) => lines.push(m),
      noGit,
    );
    expect(stamp).toBe("49dca64 2026-07-24");
    expect(lines.join("\n")).toContain("GGD_BUILD_STAMP");

    // No env, no git (the container case): warn loudly, still return a stamp.
    const warned: string[] = [];
    expect(computeBuildStamp({}, (m) => warned.push(m), noGit)).toBe(UNSTAMPED);
    expect(warned.join("\n")).toContain("BUILD STAMP MISSING");

    // A pipeline that refuses to ship an anonymous artifact opts in:
    expect(() =>
      computeBuildStamp({ [REQUIRE_STAMP_ENV_VAR]: "1" }, () => {}, noGit),
    ).toThrow(/build stamp required/);
  });

  it("utcDate is YYYY-MM-DD in UTC", () => {
    cover("build-stamp-env");
    expect(utcDate(new Date("2026-07-24T23:59:59Z"))).toBe("2026-07-24");
  });
});
